File: pump/rileylink/src/main/java/info/nightscout/androidaps/plugins/pump/common/hw/rileylink/data/RLHistoryItem.java
package info.nightscout.androidaps.plugins.pump.common.hw.rileylink.data;

import org.joda.time.LocalDateTime;

import app.aaps.core.interfaces.resources.ResourceHelper;
import info.nightscout.androidaps.plugins.pump.common.hw.rileylink.defs.RileyLinkError;
import info.nightscout.androidaps.plugins.pump.common.hw.rileylink.defs.RileyLinkServiceState;
import info.nightscout.androidaps.plugins.pump.common.hw.rileylink.defs.RileyLinkTargetDevice;
import info.nightscout.pump.common.defs.PumpDeviceState;


/**
 * Created by andy on 5/19/18.
 */

public class RLHistoryItem {

    protected LocalDateTime dateTime;
    protected RLHistoryItemSource source;
    protected RileyLinkServiceState serviceState;
    protected RileyLinkError errorCode;

    protected RileyLinkTargetDevice targetDevice;
    protected PumpDeviceState pumpDeviceState;

    public RLHistoryItem(LocalDateTime dateTime, RLHistoryItemSource source, RileyLinkTargetDevice targetDevice) {
        this.dateTime = dateTime;
        this.source = source;
        this.targetDevice = targetDevice;
    }

    public RLHistoryItem(RileyLinkServiceState serviceState, RileyLinkError errorCode,
                         RileyLinkTargetDevice targetDevice) {
        this.targetDevice = targetDevice;
        this.dateTime = new LocalDateTime();
        this.serviceState = serviceState;
        this.errorCode = errorCode;
        this.source = RLHistoryItemSource.RileyLink;
    }

    public RLHistoryItem(PumpDeviceState pumpDeviceState, RileyLinkTargetDevice targetDevice) {
        this.pumpDeviceState = pumpDeviceState;
        this.dateTime = new LocalDateTime();
        this.targetDevice = targetDevice;
        this.source = RLHistoryItemSource.MedtronicPump;
    }

    public LocalDateTime getDateTime() {
        return dateTime;
    }

    public RileyLinkServiceState getServiceState() {
        return serviceState;
    }

    public RileyLinkError getErrorCode() {
        return errorCode;
    }

    public String getDescription(ResourceHelper rh) {
        switch (this.source) {
            case RileyLink:
                return "State: " + rh.gs(serviceState.getResourceId())
                        + (this.errorCode == null ? "" : ", Error Code: " + errorCode);
            case MedtronicPump:
                return rh.gs(pumpDeviceState.getResourceId());
            default:
                return "Unknown Description";
        }
    }

    public RLHistoryItemSource getSource() {
        return source;
    }

    public PumpDeviceState getPumpDeviceState() {
        return pumpDeviceState;
    }

    public enum RLHistoryItemSource {
        RileyLink("RileyLink"), //
        MedtronicPump("Medtronic"), //
        MedtronicCommand("Medtronic"), //
        OmnipodCommand("Omnipod");

        private final String desc;


        RLHistoryItemSource(String desc) {
            this.desc = desc;
        }

        public String getDesc() {
            return desc;
        }
    }

    public static class Comparator implements java.util.Comparator<RLHistoryItem> {
        @Override
        public int compare(RLHistoryItem o1, RLHistoryItem o2) {
            return o2.dateTime.compareTo(o1.getDateTime());
        }
    }

}

File: pump/rileylink/src/main/kotlin/app/aaps/pump/common/hw/rileylink/data/RLHistoryItem.kt
package app.aaps.pump.common.hw.rileylink.data

import app.aaps.core.interfaces.pump.defs.PumpDeviceState
import app.aaps.core.interfaces.resources.ResourceHelper
import app.aaps.pump.common.extensions.stringResource
import app.aaps.pump.common.hw.rileylink.data.RLHistoryItem.RLHistoryItemSource
import app.aaps.pump.common.hw.rileylink.defs.RileyLinkError
import app.aaps.pump.common.hw.rileylink.defs.RileyLinkServiceState
import app.aaps.pump.common.hw.rileylink.defs.RileyLinkTargetDevice
import org.joda.time.LocalDateTime

/**
 * Created by andy on 5/19/18.
 */
open class RLHistoryItem(val dateTime: LocalDateTime = LocalDateTime(), val source: RLHistoryItemSource, @Suppress("unused") val targetDevice: RileyLinkTargetDevice) {

    var serviceState: RileyLinkServiceState = RileyLinkServiceState.NotStarted
    var errorCode: RileyLinkError? = null
    var pumpDeviceState: PumpDeviceState = PumpDeviceState.NeverContacted

    constructor(serviceState: RileyLinkServiceState, errorCode: RileyLinkError?, targetDevice: RileyLinkTargetDevice) : this(source = RLHistoryItemSource.RileyLink, targetDevice = targetDevice) {
        this.serviceState = serviceState
        this.errorCode = errorCode
    }

    constructor(pumpDeviceState: PumpDeviceState, targetDevice: RileyLinkTargetDevice) : this(source = RLHistoryItemSource.MedtronicPump, targetDevice = targetDevice) {
        this.pumpDeviceState = pumpDeviceState
    }

    open fun getDescription(rh: ResourceHelper): String {
        return when (this.source) {
            RLHistoryItemSource.RileyLink     -> "State: " + rh.gs(serviceState.resourceId) +
                if (this.errorCode == null) "" else ", Error Code: $errorCode"

            RLHistoryItemSource.MedtronicPump -> rh.gs(pumpDeviceState.stringResource())
            else                              -> "Unknown Description"
        }
    }

    enum class RLHistoryItemSource(val desc: String) {
        RileyLink("RileyLink"),
        MedtronicPump("Medtronic"),
        MedtronicCommand("Medtronic"),
        OmnipodCommand("Omnipod");
    }

    class Comparator : java.util.Comparator<RLHistoryItem> {

        override fun compare(o1: RLHistoryItem, o2: RLHistoryItem): Int {
            return o2.dateTime.compareTo(o1.dateTime)
        }
    }
}
